File: src/main/specnaz/src/main/java/org/specnaz/impl/TreeNode.java
package org.specnaz.impl;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

import static java.lang.String.format;

public final class TreeNode<T> {
    public final T value;
    private TreeNode<T> parent = null;
    private final List<TreeNode<T>> children = new LinkedList<>();

    public TreeNode(T value) {
        this.value = value;
    }

    public void attach(TreeNode<T> node) {
        if (node.parent != null) {
            throw new IllegalArgumentException(format("Node %s already has a parent", node));
        }
        node.parent = this;
        children.add(node);
    }

    public TreeNode<T> parent() {
        return parent;
    }

    public List<TreeNode<T>> children() {
        return Collections.unmodifiableList(children);
    }
}

File: src/main/specnaz-kotlin-old/src/main/kotlin/org/specnaz/impl/tree/TreeNode.kt
package org.specnaz.impl.tree

data class TreeNode<T>(val value: T) {
    var parent: TreeNode<T>? = null
        private set
    var children: List<TreeNode<T>> = emptyList()
        private set

    fun attach(node: TreeNode<T>) {
        if (node.parent != null)
            throw IllegalArgumentException("Cannot attach an already attached node $node")
        node.parent = this
        children += node
    }
}
